### ISSUE TITLE
 test MMS/advection/weno3 is broken

### ISSUE THREAD

**Issue by dschwoerer** [boutproject/BOUT-dev#1499]
Should be fixed or disabled/removed

https://omadesala.alfa1024.info/BEST/2019-01-15_20:01/

Related: #1049

**Comment by dschwoerer** [comment]
```
advection.cxx: In member function ‘virtual int AdvectMMS::init(bool)’:
advection.cxx:12:17: warning: unused parameter ‘restarting’ [-Wunused-parameter]
   int init(bool restarting) {
                 ^~~~~~~~~~
At global scope:
cc1plus: warning: unrecognized command line option ‘-Wno-cast-function-type’
rm: cannot remove 'data/BOUT.dmp.*.nc': No such file or directory
getmpirun: using the default mpirun -np
Making MMS advection test
  Running with method=0 mesh:ddx:upwind=w3 mesh:ddz:upwind=w3 mesh:nx=68 mesh:dx=0.09817477042468103 MZ=64
 64 |  12.657 s | ?
  -> Error norm: l-2 0.000457 l-inf 0.002288
  Running with method=0 mesh:ddx:upwind=w3 mesh:ddz:upwind=w3 mesh:nx=132 mesh:dx=0.04908738521234052 MZ=128
 128 |  19.825 s | 2.400913
  -> Error norm: l-2 0.000086 l-inf 0.000590
Convergence order = 2.400913
............ FAIL
 => Some failed tests
```

**Comment by ZedThree** [comment]
There's at least two different things going on here. One is the boundary conditions: they are currently `dirichlet_o2`. Changing this to `dirichlet_o4` surprisingly makes things worse, dropping the `l_2` order from 2.4 to 2.2 for the test. Maybe `dirichlet_o4` is not actually 4th order?

The second thing is that WENO3 is 3rd order at best and drops to lower order in regions of steep gradients. I'm not sure what counts as "steep", so I looked at the scaling of the error in the regions where the bracket term was less/more than either 1.0 or 0.5 (where the max value for this test is 3.0), and the order is mostly > 3 for the small gradient regions, except when nx > 128.

I'm not convinced about the latter though, as the function is pretty smooth. I've not found a good reference for a WENO scheme that's been MMS'd for an example function. The papers I have found have been a bit hit and miss about whether they actually get the expected order or not too.

**Comment by dschwoerer** [comment]
I tired dirichlet_nu_o4 from #1179 - and got convergence of 1.3
With dirichlet_nu_o3 I got convergence of 3.2:
```
Making MMS advection test
  Running with method=0 mesh:ddx:upwind=w3 mesh:ddz:upwind=w3 mesh:nx=68 mesh:dx=0.09817477042468103 MZ=64
 64 |  52.668 s | ?
  -> Error norm: l-2 0.000977 l-inf 0.011281
  Running with method=0 mesh:ddx:upwind=w3 mesh:ddz:upwind=w3 mesh:nx=132 mesh:dx=0.04908738521234052 MZ=128
 128 | 112.286 s | 3.194681
  -> Error norm: l-2 0.000107 l-inf 0.001463
  Running with method=0 mesh:ddx:upwind=w3 mesh:ddz:upwind=w3 mesh:nx=260 mesh:dx=0.02454369260617026 MZ=256
 256 | 177.491 s | 2.312257
  -> Error norm: l-2 0.000021 l-inf 0.000548
Convergence order = 2.312257
```
With dirichlet_nu_o4, dirichlet_o4, dirichlet_nu_o5 and dirichlet_nu_o6, I get:
```
  Running with method=0 mesh:ddx:upwind=w3 mesh:ddz:upwind=w3 mesh:nx=68 mesh:dx=0.09817477042468103 MZ=64
 64 |  32.495 s | ?
  -> Error norm: l-2 0.001121 l-inf 0.021848
  Running with method=0 mesh:ddx:upwind=w3 mesh:ddz:upwind=w3 mesh:nx=132 mesh:dx=0.04908738521234052 MZ=128
 128 | 118.946 s | 1.343173
  -> Error norm: l-2 0.000442 l-inf 0.014576
  Running with method=0 mesh:ddx:upwind=w3 mesh:ddz:upwind=w3 mesh:nx=260 mesh:dx=0.02454369260617026 MZ=256
 256 | 434.081 s | 1.322886
  -> Error norm: l-2 0.000177 l-inf 0.007466
Convergence order = 1.322886
```
As to why the o4 boundary is so much worse than o3 - I don't know.

**Comment by dschwoerer** [comment]
This is about MMS/advection/weno3 - where as #1049 is about MMS/derivative3
The advection one is using the bracket, while MMS/derivative3 is using the derivative more directly.
But if we cannot get the derivative to converge in the most simple case (#1049 doesn't need boundaries) it is not surprising this fails.

**Comment by dschwoerer** [comment]
PS: I assume you meant #1049

**Comment by ZedThree** [comment]
Fixed in #1528
